analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Globalization;
+
+public class Program
+{
+    void Constructors()
+    {
+        _ = new DateTime(1);                                                                 // Noncompliant {{Prefer using "DateTimeOffset" struct instead of "DateTime"}}
+//          ^^^^^^^^^^^^^^^
+        _ = new DateTime(1, 1, 1);                                                           // Noncompliant
+        _ = new DateTime(1, 1, 1, new GregorianCalendar());                                  // Noncompliant
+        _ = new DateTime(1, 1, 1, 1, 1, 1);                                                  // Noncompliant
+        _ = new DateTime(1, 1, 1, 1, 1, 1, new GregorianCalendar());                         // Noncompliant
+        _ = new DateTime(1, 1, 1, 1, 1, 1, DateTimeKind.Utc);                                // Noncompliant
+        _ = new DateTime(1, 1, 1, 1, 1, 1, 1);                                               // Noncompliant
+        _ = new DateTime(1, 1, 1, 1, 1, 1, 1, new GregorianCalendar());                      // Noncompliant
+        _ = new DateTime(1, 1, 1, 1, 1, 1, 1, DateTimeKind.Utc);                             // Noncompliant
+        _ = new DateTime(1, 1, 1, 1, 1, 1, 1, new GregorianCalendar(), DateTimeKind.Utc);    // Noncompliant
+    }
+
+    void Fields()
+    {
+        _ = DateTime.MaxValue;  // Noncompliant
+//          ^^^^^^^^^^^^^^^^^
+        _ = DateTime.MinValue;  // Noncompliant
+    }
+
+    void StaticProperties()
+    {
+        _ = DateTime.Now; // Noncompliant
+//          ^^^^^^^^^^^^
+        _ = DateTime.Today; // Noncompliant
+        _ = DateTime.UtcNow; // Noncompliant
+    }
+
+    void Methods(DateTime date)
+    {
+        date.Add(TimeSpan.Zero);
+        date.AddDays(0);
+        date.AddHours(0);
+        date.AddMilliseconds(0);
+        date.AddMinutes(0);
+        date.AddMonths(0);
+        date.AddSeconds(0);
+        date.AddTicks(0);
+        date.AddYears(0);
+        date.CompareTo(date);
+        DateTime.Compare(date, date);
+        DateTime.DaysInMonth(1, 1);
+        DateTime.Equals(date, date);
+        date.Equals(date);
+        DateTime.FromBinary(1);
+        DateTime.FromFileTime(1);
+        DateTime.FromFileTimeUtc(1);
+        DateTime.FromOADate(1);
+        date.GetDateTimeFormats('a');
+        date.GetHashCode();
+        date.GetTypeCode();
+        date.IsDaylightSavingTime();
+        DateTime.IsLeapYear(1);
+        DateTime.Parse("06/01/1993");                          // Noncompliant
```

Comment:
I am not 100% convinced we should raise on the methods.
Why?

- For the parse methods if the string contains only the information for a date, you cannot create a valid DateTimeOffset.
- For the specifykind the reasoning is similar, if you have a date, at this point when you are specifying the kind you might not have the offset information.

In my opinion if we only raise for constructors and the well known fields like, Now, UTCNow, Today, etc. We will ensure that the sources of DateTime objects are converted, and the change will propagate trough the whole application and everything will be changed to DateTimeOffset.

Reply:
Can you provide an example of a DateTime string that would be not parsed correctly by the DateTimeOffset.Parse() method?
I tried playing a bit with[ this example](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBfsQAwAEGECApgMoLKwDmdAvHQCIaAVgD0NYqMI0aADgEBuCgDcIyBk2YAVKGGZ86AEU069AOgAKagM7MAFIxbtOMLgEol5VesfbdzAHkAMyDbBANjFlNAkLDLG3tfZ24PCipiAE4HE39U8mos32jg0OYEDyA===) and they appear to be working similarly.

Reply:
That is exactly the problem there. Is it a conscious decision? I am afraid people will just replace DateTime with DateTimeOffset and will have always 0 as the offset. Whan in my opinion the rule should achieve is to "open the eyes" of the devs and tell them to have a conscious decision and use the offset to define where the time points to.
If the offset is always 0 because the string does not contain a valid offset, how is that DateTimeOffset objects different from a simple DateTime with Local kind?

Reply:
For DateTime.Parse the kind is Unspecified, so the information is ambiguous from the very beginning. IMO when no offset is specified it is safe to assume that we are talking about UTC, unfortunately, Microsoft does not agree:

> However, for [DateTime](https://learn.microsoft.com/en-us/dotnet/api/system.datetime) values whose [Kind](https://learn.microsoft.com/en-us/dotnet/api/system.datetime.kind) property is [DateTimeKind.Unspecified](https://learn.microsoft.com/en-us/dotnet/api/system.datetimekind#system-datetimekind-unspecified), these two conversion methods produce a [DateTimeOffset](https://learn.microsoft.com/en-us/dotnet/api/system.datetimeoffset) value whose offset is that of the local time zone.

[Source](https://learn.microsoft.com/en-us/dotnet/standard/datetime/converting-between-datetime-and-offset#conversions-from-datetime-to-datetimeoffset)

So I think I agree with you, I'll remove the parsing methods from here.